pandas/core/generic.py

```diff
@@ -2909,10 +2910,11 @@ def to_latex(self, buf=None, columns=None, col_space=None, header=True,
 
     def to_csv(self, path_or_buf=None, sep=",", na_rep='', float_format=None,
                columns=None, header=True, index=True, index_label=None,
-               mode='w', encoding=None, compression='infer', quoting=None,
-               quotechar='"', line_terminator=None, chunksize=None,
-               tupleize_cols=None, date_format=None, doublequote=True,
-               escapechar=None, decimal='.'):
+               mode='w', encoding=None,
+               compression: Union[str, Dict, None] = 'infer',
```

Comment:
Should always have types for Container classes like Dict. I think this should read `Optional[Union[str, Dict[str, str]]]`

Reply:
Right; I think it should be `Optional[Union[str, Dict[str, Optional[str]]]]`, though, as both the method and archive name can be `None`.

Reply:
What is the use case for an optional value in the dict? Isn't the point of accepting one in the first place for it to always have a `method` key with a `str` value associated with it?

Reply:
@WillAyd If there's some `compression_method`, including the `Optional` allows `method` to be `None` so we can have `compression` which is passed to `to_csv` be, for example,
```
compression = dict(method=compression_method, **other_args)
```
instead of something more cumbersome like
```
if compression_method is not None:
  compression = dict(method=compression_method, **other_args)
else:
  compression = None
```
which is admittedly not a huge difference, but passing `None` as either `method` or `archive_name` is supported regardless.